### PR TITLE
405 wen logits

### LINKED ISSUE #405: wen logits
Multiple choice task should be based on logits. This improves the task in two ways:
- Avoid brittle parsing logic when extracting miner prediction from text string
- Measure confidence of miner response

Miners should respond with a string that we can load as JSON. This will enable us to score the (re-normalized) probability assigned to the correct choice instead of a binary reward signal.

## Example
```python
miner_response = {
    "A": 1,
    "B": 0.6,
    "C": 0.25,
    "D": 0.05
}
```
Proposed reward mechanism:
```python
reward = miner_response.get( correct_response, 0 )
```
So if the  correct answer is `"C"` the miner will score `0.25` instead of `0`, which is what the current mechanism would reward.

Note that the current approach is equivalent to `miner_reward = ( argmax(miner_response) == true_index )`.

### PR DESCRIPTION
**Pull request by steffencruz**
Miners respond with logit probabilities for multiple choice tasks. Uses sparsification techniques to reduce data transferred by up to 30,000x compared to baseline approaches when using modern transformer tokenizers. Closes #405

### CODE REVIEW

**Review by Hollyqui — APPROVED** [review]
Subject to double-checking comments

File: prompting/rewards/multi_choice.py [hunk 1]
```diff
@@ -14,28 +15,61 @@ class MultiChoiceRewardModel(BaseRewardModel):
...
+            raise ValueError("Values must sum to 1")
+
+        return valid_choices
+
+    def _logit_reward(self, reference: str, completion: str) -> float:
+        """Compute difference scores given a completion and reference pair."""
+        try:
+            loaded_json = self.safe_load_json(completion)
+            valid_choices = self.process_predictions(loaded_json)
+            return valid_choices.get(reference.upper(), 0)
```
[the remark this answers is not in the document]
Comment (steffencruz) [reply prompting/rewards/multi_choice.py:58]: Oops `reference.upper()` was a mistake, should have been the loaded_json keys instead.

My logic is that we want some flexibility in handling the miner completions: e.g. `"a"` and `"A"` should both be acceptable, which allows miners to pass their raw logits with less processing. I don't feel very strongly about this though.

**Comment by steffencruz** [comment]
Could you please rebase this onto pre-staging @Hollyqui

File: prompting/rewards/multi_choice.py [hunk 2]
```diff
@@ -1,49 +1,76 @@
...
-                reward = 0
-            timings.append(time.perf_counter() - t0)
+
+        for completion in response_event.completions:
+            logger.debug(f"Completion: {completion}, reference: {reference}")
+            start_time = time.perf_counter()
+
+            reward = self.logit_reward(reference, completion)
+            if reward is None:
+                reward = self.classical_reward(reference, completion) * self.json_penalty
```
Comment (dbobrenko) [inline prompting/rewards/multi_choice.py:70]: I guess it should be `classic_reward` :)

**Review by dbobrenko — APPROVED** [review]
LGTM
